### ISSUE TITLE
Unable to start bigbluebutton-exporter service

### ISSUE THREAD

**Issue by jeanfrancoisvarin** [greenstatic/bigbluebutton-exporter#32]
Hello Gregor,

Thanks for your great job. I'm a Grafana fan (and newbie as well).

As I'm not found of Docker (I rather prefer to understand what I'm running and how it is set up), I decided to install the systemd version, following the howto step by step.

Now I'm stuck unable to start the bigbluebutton service.

When I check the status :

jfvarin@labo:/etc/bigbluebutton-exporter$ sudo systemctl status bigbluebutton-exporter
● bigbluebutton-exporter.service - BigBlueButton Exporter
   Loaded: loaded (/lib/systemd/system/bigbluebutton-exporter.service; enabled; vendor preset: enabled)
   Active: activating (auto-restart) (Result: exit-code) since lun. 2020-05-25 21:09:21 CEST; 2s ago
  Process: 17957 ExecStart=/usr/bin/python3 /opt/bigbluebutton-exporter/bbb-exporter/server.py (code=exited, status=1/FAILURE)
 Main PID: 17957 (code=exited, status=1/FAILURE)

mai 25 21:09:21 labo systemd[1]: bigbluebutton-exporter.service: Unit entered failed state.
mai 25 21:09:21 labo systemd[1]: bigbluebutton-exporter.service: Failed with result 'exit-code'.

So I went a step further by running

jfvarin@labo:/etc/bigbluebutton-exporter$ sudo /usr/bin/python3 /opt/bigbluebutton-exporter/bbb-exporter/server.py
Traceback (most recent call last):
  File "/opt/bigbluebutton-exporter/bbb-exporter/server.py", line 6, in <module>
    import settings
  File "/opt/bigbluebutton-exporter/bbb-exporter/settings.py", line 17, in <module>
    API_BASE_URL = validate_api_base_url(os.environ["API_BASE_URL"])
  File "/usr/lib/python3.5/os.py", line 725, in __getitem__
    raise KeyError(key) from None
KeyError: 'API_BASE_URL'

I triple checked the API_BASE_URL and the API_SECRET which work like a charm with the API Mate but it seems that server.py is not able to get the info.

My conf is quite standard : Ubuntu 16.04 with BBB and Greenlight + SSL + prometheus + python3-pip..,

If someone can give me a clue, I will really appreciate as I want to avoid running docker (sorry for the docker fans).

Thanks in advance for your help, I still have a lot to do to see it fully running.

Best regards from Paris.
JF

**Comment by greenstatic** [comment]
Seems to me like you didn't edit the settings properly. Did you follow [steps 4-5](https://bigbluebutton-exporter.greenstatic.dev/installation/bigbluebutton_exporter/#4-copy-systemd-unit-service-and-example-settings)?

**Comment by greenstatic** [comment]
Also, for future reference, it's much easier to read error snippets if you insert them as "code".

https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#code-and-syntax-highlighting

**Comment by greenstatic** [comment]
Just letting you know, I updated the docs regarding systemd installation with a bit safer `cp` command in step 4. It is functionally equivalent to the previous command but a bit safer if anybody mistypes anything. Just so you don't bang your head that something "looks different" than before.

**Comment by jeanfrancoisvarin** [comment]
Hello Gregor,

Thanks for your quick answer.

I did tons of experiences in order to both understand and fix the issue. 
It's good to learn new technics after 60 ;-)

I checked all Python requiered modules, checked all versions...
Re-installed with the new version (included cp) and the result stayed the same.

My conclusion was that the script was not able to load the 2 env variables so I modify the `setting.py` script by adding the `dotenv module` and addressing the `settings.env` file

ran the script both as user and sudo and worked like a charm but still unable to run it as a service...

It will probably still take some time but once it's done it's done.

I apologize not using Markdown highlighting syntax but it was first git post ever. I learn everyday.

Kind regards from France

**Comment by greenstatic** [comment]
@jeanfrancoisvarin if I will be so eager to learn new techniques at 60 I will be trilled :)

Can you report back with the output of:
```shell
$ sudo journalctl -u bigbluebutton-exporter
```

**Comment by jeanfrancoisvarin** [comment]
My pleasure.

I just think that the day I will not learn anything anymore I will be dead ;-)

I also put the modified script I tried.

JF
[journalctl2705.txt](https://github.com/greenstatic/bigbluebutton-exporter/files/4689571/journalctl2705.txt)

[settings.txt](https://github.com/greenstatic/bigbluebutton-exporter/files/4689580/settings.txt)

**Comment by greenstatic** [comment]
Ah I see the issue. Your second original snippet mislead me - since you are running it manually without having the environment variables in your session. 

You are getting in your errors:
```
mai 27 16:56:05 labo systemd[1]: Stopped BigBlueButton Exporter.
mai 27 16:56:05 labo systemd[1]: Started BigBlueButton Exporter.
mai 27 16:56:05 labo python3[25933]: Traceback (most recent call last):
mai 27 16:56:05 labo python3[25933]:   File "/opt/bigbluebutton-exporter/bbb-exporter/server.py", line 5, in <module>
mai 27 16:56:05 labo python3[25933]:     from prometheus_client import start_http_server, REGISTRY
mai 27 16:56:05 labo python3[25933]: ImportError: No module named 'prometheus_client'
mai 27 16:56:05 labo systemd[1]: bigbluebutton-exporter.service: Main process exited, code=exited, status=1/FAILURE
mai 27 16:56:05 labo systemd[1]: bigbluebutton-exporter.service: Unit entered failed state.
mai 27 16:56:05 labo systemd[1]: bigbluebutton-exporter.service: Failed with result 'exit-code'.
```

This is a library issue. The documentation is missing a vital step, installing via pip the Python dependencies. Looks like I had them already installed when I was writing the documentation.

Give this a try:
```
$ sudo apt update
$ sudo apt install python3-pip  # installs Python3 pip if you don't already have it
$ sudo pip3 install -r /opt/bigbluebutton-exporter/requirements.txt  # install Python3 dependencies system-wide
```

Report back if it works, so we can fix the documentation.

**Comment by jeanfrancoisvarin** [comment]
Thanks for the idea of using `requierements.txt` to install dependencies, 
I did it manually first and using the script allowed me to update some versions.

Anyway, it still don't grab the vars either when manually running `/usr/bin/python3 /opt/bigbluebutton-exporter/bbb-exporter/server.py` after failing `sudo systemctl start bigbluebutton-exporter`

I reboot the machine, logged in and started the service first and checked status just after.
Status was green with an error message (no module named xmltodict)...

For your information, in my previous tests, the first issue I face was xmltodict I installed manually...
THAT was my error !!!

Then turn back to grey.

Took a breath... uninstalled all xmltodict versions (pip, pip3) and reinstalled with the script.

Restarted the service, status check, netstat check and now it works ;-)

Hope for you not too much (old) people want to install it without Docker ;-)

JF

**Comment by greenstatic** [comment]
Great, I'll update the documentation shortly to reflect this.

> Hope for you not too much (old) people want to install it without Docker ;-)

No problem, that's why the instructions are there :)

**Comment by jeanfrancoisvarin** [comment]
I'll keep you posted once everything runs.

I'll be pleased to help noobiewise speaking.

JF

**Comment by jeanfrancoisvarin** [comment]
Hi Gregor,

As I told you, quick update.
Even if it's not perfect (still some issues with the metrics), my Grafana serveur is up and running grabing from my lab bbb serveur.

I just want to point out anothe mistake I did yesterday which was to run `sudo apt install prometheus` directly on the 16.04.
This version is completely obsolete and `prometheus.yml` files are not interpreted correctly.
So I moved to the latest version of Prometheus.

Again, thanks for your help.

Kind regards
JF
